test/model/internal_link_test.dart

```diff
@@ -0,0 +1,489 @@
+import 'package:zulip/model/store.dart';
+
+import '../example_data.dart' as eg;
+import 'test_store.dart';
+
+PerAccountStore setupStore({
+  required Uri realmUrl,
+  List<ZulipStream>? streams,
+  List<User>? users,
+}) {
+  final account = eg.selfAccount.copyWith(realmUrl: realmUrl);
+  final store = eg.store(account: account);
+  if (streams != null) {
+    store.addStreams(streams);
+  }
+  store.addUser(eg.selfUser);
+  if (users != null) {
+    store.addUsers(users);
+  }
+  return store;
+}
+
+void main() {
+  final realmUrl = Uri.parse('https://example.com/');
+
+  checkExpectedNarrows(List<(String, Narrow?)> testCases, {
+    List<ZulipStream>? streams,
+    PerAccountStore? store,
+    List<User>? users,
+  }) async {
+    assert((streams != null || users != null) ^ (store != null));
+    if (store == null) {
+      streams ??= [];
+      store = setupStore(realmUrl: realmUrl, streams: streams, users: users);
+    }
+    for (final testCase in testCases) {
+      final String path = testCase.$1;
+      final Uri url = realmUrl.resolve(path);
+      final Narrow? expected = testCase.$2;
+      check(because: path, parseInternalLink(url, store)).equals(expected);
+    }
+  }
+
+  group('Url acceptance', () {
+    final streams = [
+      eg.stream(streamId: 1, name: 'check'),
+    ];
+    final testCases = [
+      (true, 'legacy: stream name, no ID',
+        realmUrl.resolve('#narrow/stream/check'), realmUrl),
+      (true, 'legacy: stream name, no ID, topic',
+        realmUrl.resolve('#narrow/stream/check/topic/topic1'), realmUrl),
+
+      (true, 'with numeric stream ID',
+        realmUrl.resolve('#narrow/stream/123-check'), realmUrl),
+      (true, 'with numeric stream ID and topic',
+        realmUrl.resolve('#narrow/stream/123-check/topic/topic1'), realmUrl),
+
+      (true, 'with numeric pm user IDs (new operator)',
+        realmUrl.resolve('#narrow/dm/123-mark'), realmUrl),
+      (true, 'with numeric pm user IDs (old operator)',
+        realmUrl.resolve('#narrow/pm-with/123-mark'), realmUrl),
+
+      (false, 'wrong fragment',
+        realmUrl.resolve('#nope'), realmUrl),
+      (false, 'wrong path',
+        realmUrl.resolve('user_uploads/#narrow/stream/check'), realmUrl),
+      (false, 'wrong domain',
+        Uri.parse('https://another.com/#narrow/stream/check'), realmUrl),
+
+      (false, '#narrowly',
+        realmUrl.resolve('#narrowly/stream/check'), realmUrl),
+
+      (false, 'double slash',
+        Uri.parse('https://example.com//#narrow/stream/check'), realmUrl),
+      (false, 'triple slash',
+        Uri.parse('https://example.com///#narrow/stream/check'), realmUrl),
+
+      (true, 'with port',
+        Uri.parse('https://example.com:444/#narrow/stream/check'),
+        Uri.parse('https://example.com:444/')),
+
+      // In the following examples unicode domains will
+      // pass but punycode variants of unicode domains are not
+      // treated as matching because Dart's [Uri] currently
```

Comment:
This description "In the following … matching" isn't quite right. Punycode is involved only in the first of the three failing examples; and it's not a variant, but an encoding.

I think the description can actually just be left out, though — the examples speak for themselves about what doesn't work. What this comment can add is the next few lines, which point to why they don't work.

(cf https://github.com/zulip/zulip-flutter/pull/305#discussion_r1336378571 )